src/cosalib/qemuvariants.py

```diff
@@ -187,22 +200,26 @@ def mutate_image(self, callback=None):
             callback(work_img)
 
         if self.tar_members:
-            wmode = 'w'
-            if final_img.endswith('gz'):
-                wmode = 'w:gz'
-            elif final_img.endswith('xz'):
-                wmode = 'w:xz'
-            log.info(f"Preparing tarball with mode '{wmode}': {final_img}")
-            with tarfile.open(final_img, wmode) as tar:
-                base_disk = self.tar_members.pop(0)
-                base_name = os.path.basename(base_disk)
-                log.info(f" - adding base disk '{work_img}' as '{base_name}'")
-                tar.add(work_img, arcname=base_name)
-
-                for te in self.tar_members:
-                    te_name = os.path.basename(te)
-                    log.info(f" - adding additional file: {te_name}")
-                    tar.add(te, arcname=te_name)
+            # Python does not create sparse Tarfiles, so we have do it via
+            # the CLI here.
+            log.info(f"Preparing tarball {final_img} with flags '{self.tar_flags}'")
+
+            base_name = self.tar_members.pop(0)
+            tarlist = [base_name]
+            if base_name != os.path.basename(work_img):
+                copy_sparse(work_img, os.path.join(self._tmpdir, base_name))
```

Comment:
I don't understand the circumstances in which this condition fires.   Is it something like enforcing a naming scheme for the file *inside* the tarball?  Wouldn't it be a lot cheaper to just make a symlink in that case, since we're passing `h` to the tar invocation?

Actually at a higher level, it seems like it'd be a lot simpler to me if the builder implementations that used `tar_files` just each directly called `tar` as they need to.

Today only gcp and vmware use this AFAICS.

Reply:
This  PR fixes the issue today. I can look at refactoring this out later. The original reason why I put the Tarball functionality here was because of the distaste for OOO.

Reply:
Also, the latest push includes a comment about when the condition would happen.

Reply:
Hmm, I don't think we need `cp --sparse=always` anymore with https://github.com/coreos/coreos-assembler/pull/1097. The output of `qemu-img convert` is already sparse. Seems like here we just want `os.link()` if it's to have the filenames match up? That makes for less IO too. :)